internal/core/src/index/NgramInvertedIndex.cpp

```diff
@@ -0,0 +1,152 @@
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+
+#include "index/NgramInvertedIndex.h"
+#include "exec/expression/Expr.h"
+
+namespace milvus::index {
+NgramInvertedIndex::NgramInvertedIndex(const storage::FileManagerContext& ctx,
+                                       const NgramParams& params)
+    : min_gram_(params.min_gram), max_gram_(params.max_gram) {
+    schema_ = ctx.fieldDataMeta.field_schema;
+    field_id_ = ctx.fieldDataMeta.field_id;
+    mem_file_manager_ = std::make_shared<MemFileManager>(ctx);
+    disk_file_manager_ = std::make_shared<DiskFileManager>(ctx);
+
+    if (params.loading_index) {
+        path_ = disk_file_manager_->GetLocalNgramIndexPrefix();
+    } else {
+        path_ = disk_file_manager_->GetLocalTempNgramIndexPrefix();
+        boost::filesystem::create_directories(path_);
+        d_type_ = TantivyDataType::Keyword;
+        std::string field_name =
+            std::to_string(disk_file_manager_->GetFieldDataMeta().field_id);
+        wrapper_ = std::make_shared<TantivyIndexWrapper>(
+            field_name.c_str(), path_.c_str(), min_gram_, max_gram_);
+    }
+}
+
+void
+NgramInvertedIndex::BuildWithFieldData(const std::vector<FieldDataPtr>& datas) {
+    AssertInfo(schema_.data_type() == proto::schema::DataType::String ||
+                   schema_.data_type() == proto::schema::DataType::VarChar,
+               "schema data type is {}",
+               schema_.data_type());
+    index_build_begin_ = std::chrono::system_clock::now();
+    InvertedIndexTantivy<std::string>::BuildWithFieldData(datas);
+}
+
+IndexStatsPtr
+NgramInvertedIndex::Upload(const Config& config) {
+    finish();
+    auto index_build_end = std::chrono::system_clock::now();
+    auto index_build_duration =
+        std::chrono::duration<double>(index_build_end - index_build_begin_)
+            .count();
+    LOG_INFO("index build done for ngram index, field id: {}, duration: {}s",
+             field_id_,
+             index_build_duration);
+    return InvertedIndexTantivy<std::string>::Upload(config);
+}
+
+void
+NgramInvertedIndex::Load(milvus::tracer::TraceContext ctx,
+                         const Config& config) {
+    auto index_files =
+        GetValueFromConfig<std::vector<std::string>>(config, INDEX_FILES);
+    AssertInfo(index_files.has_value(),
+               "index file paths is empty when load ngram index");
+
+    auto files_value = index_files.value();
+    auto it = std::find_if(
+        files_value.begin(), files_value.end(), [](const std::string& file) {
+            return file.substr(file.find_last_of('/') + 1) ==
```

Comment:
instead of performing a find_last_of and a substr, we can: `file.size() >= suffix.size() && std::equal(suffix.rbegin(), suffix.rend(), file.rbegin())` where the suffix is `"/index_null_offset"`.